### ISSUE TITLE
A full no_std implementation.

### ISSUE THREAD

**Issue by darleybarreto** [x-hgg-x/tz-rs#38]
Hi there,

Would you be interested in merging a contribution to make `tz-rs` full no_std? 

I am asking this because I think this crate can help [relibc](https://gitlab.redox-os.org/redox-os/relibc) a lot. For both Redox and Linux, we basically have all the necessary building blocks to use `tz-rs`, including logging etc. 

What would be the best approach here? Having traits for everything that  requires `std` and then implementing on `relibc`?  And for those who can use std, nothing changes, just need to add boilerplate to current code.

Kind regards,

Darley

**Comment by x-hgg-x** [comment]
> Would you be interested in merging a contribution to make tz-rs full no_std?

Yes this could be a good addition to the library.

The following methods are missing when the `std` feature is disabled :

* [TimeZone::local](https://docs.rs/tz-rs/0.6.14/tz/struct.TimeZone.html#method.local),  [TimeZone::from_tz_data](https://docs.rs/tz-rs/0.6.14/tz/struct.TimeZone.html#method.from_tz_data) and [TimeZone::from_posix_tz](https://docs.rs/tz-rs/0.6.14/tz/struct.TimeZone.html#method.from_posix_tz): depends on `std::fs` and `std::io` for parsing.
* [TimeZone::find_current_local_time_type](https://docs.rs/tz-rs/0.6.14/tz/struct.TimeZone.html#method.find_current_local_time_type) and [DateTime::now](https://docs.rs/tz-rs/0.6.14/tz/struct.DateTime.html#method.now): depends on `std::time`.

> What would be the best approach here? Having traits for everything that requires std and then implementing on relibc? And for those who can use std, nothing changes, just need to add boilerplate to current code.

A trait is a good idea, like in the following code:

```rust
pub trait StdImpl {
    fn now() -> i64;
    // ...
    // define all needed methods
}

pub struct Std;

impl StdTrait for Std {
   // impl using std
}

pub struct GenericTimeZone<T: StdImpl> {
    // ...
    std: T
}

pub type TimeZone = GenericTimeZone<Std>;
```

Adding traits methods afterwards will be a breaking change, so we should be careful when selecting needed methods.

**Comment by darleybarreto** [comment]
I see, thanks for the great feedback. I will give this a try.

**Comment by darleybarreto** [comment]
Hey, I'm not really a Rust expert, would you be able to expand more on your first comment? The pain points here are mostly the types directly used inside, so I'm having a hard time figuring this out properly.

**Comment by x-hgg-x** [comment]
This is more complicated than it seems, I will try to do a first implementation myself.

**Comment by x-hgg-x** [comment]
Implemented in `tz-rs v0.7.0`.

**Comment by darleybarreto** [comment]
Wow, thank you soo much!!
